Add sensor platform to MadVR

iloveicedgreentea opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This add sensors + tests and options flow for madvr


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:   https://github.com/home-assistant/home-assistant.io/pull/33896

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

iloveicedgreentea committed:
feat: add sensors

iloveicedgreentea committed:
feat: add tests for sensors

iloveicedgreentea committed:
feat: add options flow

iloveicedgreentea committed:
feat: add tests for options flow

iloveicedgreentea committed:
fix: remove options flow

joostlek commented on homeassistant/components/madvr/sensor.py:
```diff
@@ -0,0 +1,239 @@
+
+
+SENSORS: tuple[MadvrSensorEntityDescription, ...] = (
+    MadvrSensorEntityDescription(
+        key="mac_address",
+        name="MAC Address",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        value_fn=lambda coordinator: coordinator.data.get("mac_address"),
+        translation_key="mac_address",
+    ),
```
We already have this in the device info, no need for separate sensor

joostlek commented on homeassistant/components/madvr/sensor.py:
```diff
@@ -0,0 +1,239 @@
+        value_fn=lambda coordinator: coordinator.data.get("incoming_res"),
+        translation_key="incoming_res",
+    ),
+    MadvrSensorEntityDescription(
+        key="incoming_frame_rate",
+        name="Incoming Frame Rate",
+        value_fn=lambda coordinator: coordinator.data.get("incoming_frame_rate"),
+        translation_key="incoming_frame_rate",
+    ),
```
We can add a frequency here

iloveicedgreentea replied:
It reports it as frame rate not refresh rate - 23.976p

I can convert it, but seems hacky IMO because 23.976p has a specific connotation besides refresh rate (its a progressive signal)

joostlek commented on homeassistant/components/madvr/sensor.py:
```diff
@@ -0,0 +1,239 @@
+        value_fn=lambda coordinator: coordinator.data.get("incoming_frame_rate"),
+        translation_key="incoming_frame_rate",
+    ),
+    MadvrSensorEntityDescription(
+        key="incoming_color_space",
+        name="Incoming Color Space",
+        value_fn=lambda coordinator: coordinator.data.get("incoming_color_space"),
+        translation_key="incoming_color_space",
+    ),
+    MadvrSensorEntityDescription(
+        key="incoming_bit_depth",
+        name="Incoming Bit Depth",
+        value_fn=lambda coordinator: coordinator.data.get("incoming_bit_depth"),
+        translation_key="incoming_bit_depth",
+    ),
+    MadvrSensorEntityDescription(
+        key="incoming_colorimetry",
+        name="Incoming Colorimetry",
+        value_fn=lambda coordinator: coordinator.data.get("incoming_colorimetry"),
+        translation_key="incoming_colorimetry",
+    ),
+    MadvrSensorEntityDescription(
+        key="incoming_black_levels",
+        name="Incoming Black Levels",
+        value_fn=lambda coordinator: coordinator.data.get("incoming_black_levels"),
+        translation_key="incoming_black_levels",
+    ),
```
What are the possible values

iloveicedgreentea replied:
in order
1. 422, 444, 420, RGB
2. 10bit, 12bit, 8bit
3. 601|PAL|709|DCI|2020
4. PC | TV

joostlek replied:
If these are all pre known we should use enum sensors to allow users to be able to preselect them in the automation editor and know what the possible values are

joostlek commented on homeassistant/components/madvr/sensor.py:
```diff
@@ -0,0 +1,239 @@
+    def native_value(self) -> float | str | None:
+        """Return the state of the sensor."""
+        return self.entity_description.value_fn(self.coordinator)
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self.native_value is not None
```
if native_value is `None` it is rendered as `unknown` and that is a good thing imo.

joostlek commented on homeassistant/components/madvr/sensor.py:
```diff
@@ -0,0 +1,239 @@
+    ),
+    MadvrSensorEntityDescription(
+        key="outgoing_bit_depth",
+        name="Outgoing Bit Depth",
```
Please remove all the names to make use of the translations

joostlek commented on tests/components/madvr/test_sensors.py:
```diff
@@ -0,0 +1,264 @@
+    await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)
+
+
+@pytest.mark.parametrize(
```
In this case, might as well just create one big data update and call this before the snapshot and make the snapshot snapshot all the updated states

joostlek replied:
In the previous test

joostlek reviewed (changes requested):
Please revert the change in config_flow.py

iloveicedgreentea committed:
fix: remove names and mac sensor, add incoming signal type

frenck commented:
Marked `docs-missing` as the linked PR belongs to a different parent PR. This PR does not have documentation.

iloveicedgreentea committed:
feat: add enum types to supported sensors

iloveicedgreentea committed:
fix: consolidate tests into snapshot

iloveicedgreentea commented:
Updated sensors to use enums and other classes, consolidate tests into snapshot, updated the docs PR to reference this one

https://github.com/home-assistant/home-assistant.io/pull/33659

iloveicedgreentea committed:
fix: use consts

joostlek commented on homeassistant/components/madvr/strings.json:
```diff
@@ -38,23 +38,20 @@
       },
       "temp_mainboard": {
-        "name": "Mainboard temperature"
+        "name": "Temp mainboard"
```
I liked the old name better since it was more readable. `GPU temperature` looks better imo

joostlek commented on tests/components/madvr/test_sensors.py:
```diff
@@ -0,0 +1,86 @@
+    update_callback(update_data)
+    await hass.async_block_till_done()
+
+    # Snapshot all entity states
+    assert snapshot == {
+        entity_id: hass.states.get(entity_id)
+        for entity_id in hass.states.async_entity_ids("sensor")
+        if entity_id.startswith("sensor.madvr_envy_")
+    }
```
Please use the snapshot platform helper in this case

iloveicedgreentea committed:
fix: update names and use snapshot platform

iloveicedgreentea commented:
updated names and snapshot platform

iloveicedgreentea committed:
fix: fix test name for new translations

iloveicedgreentea committed:
fix: comment

iloveicedgreentea committed:
fix: improve sensor names

iloveicedgreentea commented:
@joostlek I updated the docs link btw should be good now

joostlek commented on tests/components/madvr/test_sensors.py:
```diff
@@ -0,0 +1,82 @@
+    # Test invalid temperature value
+    update_callback({"temp_gpu": -1})
+    await hass.async_block_till_done()
+    assert hass.states.get("sensor.madvr_envy_gpu_temperature").state == "unknown"
```
```suggestion
    assert hass.states.get("sensor.madvr_envy_gpu_temperature").state == STATE_UNKNOWN
```

joostlek commented on tests/components/madvr/test_sensors.py:
```diff
@@ -0,0 +1,82 @@
+    # Test sensor becomes known again
+    update_callback({"incoming_res": "1920x1080"})
+    await hass.async_block_till_done()
+    assert hass.states.get("sensor.madvr_envy_incoming_resolution").state == "1920x1080"
```
Just a thought, would it be maybe more useful to return width and height in a separate sensor? Like I am thinking of use cases for this entity, and I can only thing about a display use case or when the values are separated

iloveicedgreentea replied:
this value is commonly used to match to a know set of resolutions so it would not be useful if separate

iloveicedgreentea commented on homeassistant/components/madvr/manifest.json:
```diff
@@ -6,5 +6,5 @@
   "integration_type": "device",
   "iot_class": "local_push",
-  "requirements": ["py-madvr2==1.6.27"]
+  "requirements": ["py-madvr2==1.6.28"]
```
no its required to fix how one of the sensors works

joostlek replied:
No but please bump this in a PR before this PR

iloveicedgreentea replied:
https://github.com/home-assistant/core/pull/122275

iloveicedgreentea commented on homeassistant/components/madvr/sensor.py:
```diff
@@ -0,0 +1,263 @@
+def get_temperature(coordinator: MadVRCoordinator, key: str) -> float | None:
+    """Get temperature value if valid, otherwise return None."""
+    temp = coordinator.data.get(key)
+    return float(str(temp)) if is_valid_temperature(temp) else None
```
fixed

iloveicedgreentea commented on homeassistant/components/madvr/config_flow.py:
its not related but makes the code a lot easier to work with

joostlek replied:
But it should go in a separate PR as its outside of the scope of the sensor platform

iloveicedgreentea replied:
reverted

iloveicedgreentea committed:
fix: address comments

iloveicedgreentea committed:
Merge branch 'dev' into madvr_more-entities

iloveicedgreentea committed:
feat: disable uncommon sensors by default

iloveicedgreentea committed:
Merge branch 'dev' into madvr_more-entities

iloveicedgreentea commented:
Updated doc PR: https://github.com/home-assistant/home-assistant.io/pull/33896

iloveicedgreentea committed:
fix: update sensors

iloveicedgreentea committed:
fix: revert config_flow change

iloveicedgreentea commented:
Addressed comments and updated snapshots
